proxy/src/control/cache.rs

```diff
@@ -93,17 +112,23 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
         self.reset_on_next_modification = false;
     }
 
-    pub fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
-        self.retain(&HashSet::new(), on_change)
+    pub fn clear<F>(&mut self, on_change: &mut F)
+    where
+        F: FnMut((K, V), CacheChange),
+    {
+        self.retain(&IndexMap::new(), on_change)
     }
 
-    pub fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
-        where F: FnMut(T, CacheChange)
+    pub fn retain<F, Q>(&mut self, to_retain: &IndexMap<K, V>, mut on_change: F)
```

Comment:
This should be renamed to `retain_replacing_duplicates()` and it needs to do the same kind of label comparison as `extend_replacing_duplicates()`.